src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_BasicDrillerBase.java

```diff
@@ -0,0 +1,138 @@
+    boolean isPickingPipes;
+    boolean waitMiningPipe = true;
+
+    public GT_MetaTileEntity_BasicDrillerBase(int aID, String aName, String aNameRegional, int aTier, String[] aDescription, int aInSlots, int aOutSlots, String aGUIName, ITexture... aTextures){
+        super(aID,aName,aNameRegional,aTier,1,aDescription,aInSlots,aOutSlots,aGUIName,"",aTextures);
+    }
+
+    public GT_MetaTileEntity_BasicDrillerBase(String aName, int aTier, String[] aDescription, ITexture[][][] aTextures, int aInputSlotCount, int aOutputSlotCount, String aGUIName) {
+        super(aName,aTier,1,aDescription,aTextures,aInputSlotCount,aOutputSlotCount,aGUIName,"");
+    }
+
+    public boolean allowPutStack(IGregTechTileEntity aBaseMetaTileEntity, int aIndex, byte aSide, ItemStack aStack) {
+        return (super.allowPutStack(aBaseMetaTileEntity, aIndex, aSide, aStack)) && (aStack.getItem() == MINING_PIPE.getItem());
+    }
+
+    public abstract boolean hasFreeSpace();
+
+    @Override
+    public void onPostTick(IGregTechTileEntity aBaseMetaTileEntity, long aTick) {
+        super.onPostTick(aBaseMetaTileEntity, aTick);
+        if (aBaseMetaTileEntity.isServerSide()) {
+            if (aBaseMetaTileEntity.isAllowedToWork() && aBaseMetaTileEntity.isUniversalEnergyStored(ENERGY[mTier] * (SPEED[mTier] - mProgresstime)) && hasFreeSpace()) {
+                miningPipe:
+                if (waitMiningPipe) {
+                    mMaxProgresstime = 0;
+                    for (int i = 0; i < mInputSlotCount; i++) {
+                        ItemStack s = getInputAt(i);
+                        if (s != null && s.getItem() == MINING_PIPE.getItem() && s.stackSize > 0) {
+                            waitMiningPipe = false;
+                            break miningPipe;
+                        }
+                    }
+                    return;
+                }
+                aBaseMetaTileEntity.decreaseStoredEnergyUnits(ENERGY[mTier], true);
+                mMaxProgresstime = SPEED[mTier];
+            } else {
+                mMaxProgresstime = 0;
+                return;
+            }
+            if (mProgresstime == SPEED[mTier] - 1) {
+                if (isPickingPipes) {
+                    if (drillY == 0) {
+                        aBaseMetaTileEntity.disableWorking();
+                        isPickingPipes = false;
+                    } else if (aBaseMetaTileEntity.getBlockOffset(0, drillY, 0) == MINING_PIPE_TIP_BLOCK || aBaseMetaTileEntity.getBlockOffset(0, drillY, 0) == MINING_PIPE_BLOCK) {
+                        mOutputItems[0] = MINING_PIPE.copy();
+                        mOutputItems[0].stackSize = 1;
+                        aBaseMetaTileEntity.getWorld().setBlockToAir(aBaseMetaTileEntity.getXCoord(), aBaseMetaTileEntity.getYCoord() + drillY, aBaseMetaTileEntity.getZCoord());
+                        drillY++;
+                    }
+                    return;
+                }
+                if (drillY == 0) {
```

Comment:
Выражение _drillY == 0 || drillZ > RADIUS[mTier]_ тоже стоит вынести в отдельную функцию, но название для меня вообще не очевидно, поэтому на ваше усмотрение.
```suggestion
                if (drillY == 0 || drillZ > RADIUS[mTier]) {
                    moveOneDown(aBaseMetaTileEntity);
                } else {
                    while (drillZ <= RADIUS[mTier]) {
                        while (drillX <= RADIUS[mTier]) {
                            if(workBlock(aBaseMetaTileEntity))
                              return;
                           drillX++;
                        }
                        drillX = -RADIUS[mTier];
                        drillZ++;
                    }
                }                
```

Reply:
эта проверка совершенно лишняя, помпа не может добывать жидкости на том же уровне на котором находится контроллер

Reply:
Объединил оригинальные 80 и 84 строки, чтобы избавиться от return'ов.